src/picom.c

```diff
@@ -1158,14 +1158,41 @@ void root_damaged(session_t *ps) {
 		}
 		auto pixmap = x_get_root_back_pixmap(&ps->c, ps->atoms);
 		if (pixmap != XCB_NONE) {
+			xcb_get_geometry_reply_t *r = xcb_get_geometry_reply(
+			    ps->c.c, xcb_get_geometry(ps->c.c, pixmap), NULL);
+			if (!r) {
+				goto err;
+			}
+
+			// If the root background pixmap's depth matches the root window's
+			// depth, use the root window's visual. Otherwise, find a suitable
+			// visual for the root background pixmap's depth and use it. They
+			// may not match if a root background pixmap atom points to a
+			// pixmap that's not owned by the root window, e.g. in case of the
+			// Xfce's desktop manager xfdesktop that sets the _XROOTPMAP_ID
+			// atom to a pixmap owned by it that seems to always have 32 bpp
+			// depth when the common root window's depth is 24 bpp.
```

Comment:
hmm, i feel this doesn't quite get to the point (the comment below is good though). i suggest something like:

> X acquires the root background pixmap has a depth that matches the root window, and _XROOTPMAP_ID is supposed to point to the root pixmap. So we assumed binding root pixmap with root window's visual will always succeed. But some applications don't play nice, for example: .........

Reply:
is it better now?